<a href="https://colab.research.google.com/github/Jialan8020/Assign8/blob/master/question4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Fit a multilayer neural network model on the SHHS data to predict whether rdi4p > 7 using the variables from the previous HW assigment. Compare the test set accuracy with the penalized model from your previous HW.

In [0]:
import pandas as pd
import tensorflow as tf
import numpy as np
from sklearn.model_selection import cross_val_score
from tensorflow import keras
from sklearn.model_selection import train_test_split

In [130]:
dat = pd.read_csv("shhs.txt",sep="\t")
dat.head(4)

pptid  waist  COPD15  ASTHMA15  ...  age_s1  smokstat_s1  WASO     bmi_s1
0      1   86.0     0.0       0.0  ...      55          2.0  65.0  21.777553
1      2  107.0     0.0       0.0  ...      78          0.0  43.0  32.950680
2      3   82.0     0.0       0.0  ...      77          0.0  73.0  24.114150
3      4   85.0     0.0       0.0  ...      48          0.0  43.5  20.185185

[4 rows x 30 columns]

In [144]:
dat_use = dat[["rdi4p","waist","gender","age_s1","bmi_s1"]]
dat_use = dat_use.dropna()
dat_use

rdi4p  waist  gender  age_s1     bmi_s1
0      1.438083   86.0       1      55  21.777553
1     17.802198  107.0       1      78  32.950680
2      4.853556   82.0       0      77  24.114150
3      0.797342   85.0       1      48  20.185185
4      2.756757   76.0       0      66  23.309053
...         ...    ...     ...     ...        ...
5799  34.415781   99.0       1      66  25.693134
5800  13.653603   99.0       1      54  26.730372
5801   2.163462   91.0       1      55  24.228571
5802   1.038319   93.0       1      53  25.374483
5803  29.075630  112.0       1      54  33.934559

[5416 rows x 5 columns]

In [163]:
# Extract the rdi4p column
rdi = pd.DataFrame(dat_use["rdi4p"])
rdi

rdi4p
0      1.438083
1     17.802198
2      4.853556
3      0.797342
4      2.756757
...         ...
5799  34.415781
5800  13.653603
5801   2.163462
5802   1.038319
5803  29.075630

[5416 rows x 1 columns]

In [164]:
rdi[rdi < 7] = 0
rdi[rdi > 7] = 1
rdi = rdi.astype(int)
rdi


rdi4p
0         0
1         1
2         0
3         0
4         0
...     ...
5799      1
5800      1
5801      0
5802      0
5803      1

[5416 rows x 1 columns]

In [159]:
X = dat_use[["waist","gender","age_s1","bmi_s1"]]
X

waist  gender  age_s1     bmi_s1
0      86.0       1      55  21.777553
1     107.0       1      78  32.950680
2      82.0       0      77  24.114150
3      85.0       1      48  20.185185
4      76.0       0      66  23.309053
...     ...     ...     ...        ...
5799   99.0       1      66  25.693134
5800   99.0       1      54  26.730372
5801   91.0       1      55  24.228571
5802   93.0       1      53  25.374483
5803  112.0       1      54  33.934559

[5416 rows x 4 columns]

In [0]:
# Separate training and test data
X_train, X_test, y_train, y_test = train_test_split(X, rdi, test_size=0.2)

In [0]:
model = tf.keras.Sequential([
  tf.keras.layers.Dense(64, activation='relu',input_dim=4),
  tf.keras.layers.Dense(10, activation='relu'),
  tf.keras.layers.Dense(1,activation='sigmoid')
])
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['mse'])

In [166]:
model.fit(X_train, y_train, epochs=5)

Epoch 1/5
136/136 [==============================] - 0s 1ms/step - loss: 0.6980 - mse: 0.2480
Epoch 2/5
136/136 [==============================] - 0s 1ms/step - loss: 0.6588 - mse: 0.2326
Epoch 3/5
136/136 [==============================] - 0s 1ms/step - loss: 0.6724 - mse: 0.2374
Epoch 4/5
136/136 [==============================] - 0s 1ms/step - loss: 0.6518 - mse: 0.2292
Epoch 5/5
136/136 [==============================] - 0s 1ms/step - loss: 0.6496 - mse: 0.2283


In [167]:
model.evaluate(X_test,y_test,verbose=2)

34/34 - 0s - loss: 0.6422 - mse: 0.2250


[0.6421515941619873, 0.22500526905059814]

**Comparison with HW 7**

The best cross validated error of Lasso regression is 0.8683.

The best cross validated error of Ridge regression is 0.8694.

The best cross validated error of elastic net is 0.8687.

The best cross validated error of no penalty is 0.8691.

However, by using neural network algorithm, we achieved a mse of 0.2250, which is significantly smaller than the logistic regression model.